WcaOnRails/app/webpacker/components/CompetitionsOverview/filterUtils.js

```diff
@@ -1,19 +1,105 @@
+const DEFAULT_SEARCH = '';
+const DEFAULT_EVENTS = [];
+const INCLUDE_CANCELLED_TRUE = 'on';
+
+// without time string, date will be interpreted in user's time zone
+const parseDate = (dateString) => new Date(`${dateString}T00:00:00.000`);
+const formatDate = (date) => DateTime.fromJSDate(date).toFormat('yyyy-MM-dd');
+
+export const getDisplayMode = (searchParams) => (
+  searchParams.get(DISPLAY_MODE) || DEFAULT_DISPLAY_MODE
+);
+
+export const createFilterState = (searchParams) => ({
+  timeOrder: searchParams.get(TIME_ORDER) || DEFAULT_TIME_ORDER,
+  selectedYear: searchParams.get(YEAR) || DEFAULT_YEAR,
+  customStartDate:
+    searchParams.get(START_DATE) ? parseDate(searchParams.get(START_DATE)) : DEFAULT_DATE,
+  customEndDate:
+    searchParams.get(END_DATE) ? parseDate(searchParams.get(END_DATE)) : DEFAULT_DATE,
+  region: searchParams.get(REGION) || DEFAULT_REGION,
+  delegate: searchParams.get(DELEGATE) || DEFAULT_DELEGATE,
+  search: searchParams.get(SEARCH) || DEFAULT_SEARCH,
+  selectedEvents:
+    searchParams.get(SELECTED_EVENTS)?.split(',')?.filter(Boolean) || DEFAULT_EVENTS,
+  shouldIncludeCancelled: searchParams.get(INCLUDE_CANCELLED) === INCLUDE_CANCELLED_TRUE,
+});
+
+export const updateSearchParams = (searchParams, filterState, displayMode) => {
+  const {
+    timeOrder,
+    selectedYear,
+    customStartDate,
+    customEndDate,
+    region,
+    delegate,
+    search,
+    selectedEvents,
+    shouldIncludeCancelled,
+  } = filterState;
 
-export const filterInitialState = {
-  timeOrder: 'present',
-  selectedYear: 'all_years',
-  customStartDate: null,
-  customEndDate: null,
-  region: 'all_regions',
-  delegate: '',
-  search: '',
-  selectedEvents: [],
-  shouldIncludeCancelled: false,
+  searchParams.set(DISPLAY_MODE, displayMode);
+  searchParams.delete(DISPLAY_MODE, DEFAULT_DISPLAY_MODE);
+
+  searchParams.set(TIME_ORDER, timeOrder);
+  searchParams.delete(TIME_ORDER, DEFAULT_TIME_ORDER);
+  searchParams.set(YEAR, selectedYear);
+  searchParams.delete(YEAR, DEFAULT_YEAR);
+  if (customStartDate) {
+    searchParams.set(START_DATE, formatDate(customStartDate));
+  } else {
+    searchParams.delete(START_DATE);
+  }
+  if (customEndDate) {
+    searchParams.set(END_DATE, formatDate(customEndDate));
+  } else {
+    searchParams.delete(END_DATE);
+  }
+  searchParams.set(REGION, region);
+  searchParams.delete(REGION, DEFAULT_REGION);
+  searchParams.set(DELEGATE, delegate);
+  searchParams.delete(DELEGATE, DEFAULT_DELEGATE);
+  searchParams.set(SEARCH, search);
+  searchParams.delete(SEARCH, DEFAULT_SEARCH);
+  searchParams.set(SELECTED_EVENTS, selectedEvents.join(','));
+  searchParams.delete(SELECTED_EVENTS, DEFAULT_EVENTS.join(','));
+  if (shouldIncludeCancelled) {
+    searchParams.set(INCLUDE_CANCELLED, INCLUDE_CANCELLED_TRUE);
+  } else {
+    searchParams.delete(INCLUDE_CANCELLED);
+  }
+
+  window.history.replaceState({}, '', `${window.location.pathname}?${searchParams}`);
 };
```

Comment:
Please add some line breaks to this block and reorganize it a bit:
- First, all the "primitive" valuesthat follow a simple two-line set-delete pattern. Include a comment about what this pattern does.
- Then all the values that are wrapped inside an `if`.
- Lastly, why do you use `window.history.replaceState` instead of just overriding `window.location.search`?

Reply:
I use `window.history.replaceState` because I was following what was done in the incidents log. I don't remember why I did it this way there, might have just been the first thing I found that worked, or possibly your suggested way didn't work for some reason.

Reply:
Okay, for now let's assume that you had a good reason to do it this way in the incidents log and trust your past self :)